rpc: getorphantxs follow-up

tdb3 opened the pull request:
Implements follow-up suggestions from #30793.

- Now disallows undefined verbosity levels (below and above valid values) (https://github.com/bitcoin/bitcoin/pull/30793#discussion_r1786093549)
- Disallows boolean verbosity (https://github.com/bitcoin/bitcoin/pull/30793#discussion_r1788273274) and adds guidance to developer-notes
- Checks that `getorphantxs` is a hidden rpc (https://github.com/bitcoin/bitcoin/pull/30793#discussion_r1786107786)
- Adds a test for `expiration` time
- Adds `entry` time to the returned orphan objects (verbosity >=1) to relieve the user from having to calculate it from `expiration`.  Also adds associated test. (https://github.com/bitcoin/bitcoin/pull/30793#discussion_r1743687732)
- Minor cleanup (blank line removal and log message move) (https://github.com/bitcoin/bitcoin/pull/30793#discussion_r1786092641)

Included a commit to rename the test to a more generic `get_orphans` to better accommodate future orphanage-related RPCs (e.g. `getorphanangeinfo`).  Can drop the refactor commit from this PR if people feel strongly about it.

tdb3 committed:
rpc: disallow undefined verbosity in getorphantxs

tdb3 committed:
refactor: rename rpc_getorphantxs to rpc_orphans

Generalizes the test to accommodate additional
orphan-related RPCs

tdb3 committed:
rpc: add entry time to getorphantxs

tdb3 committed:
test: add entry and expiration time checks

tdb3 committed:
test: check that getorphantxs is hidden

tdb3 commented on test/functional/rpc_orphans.py:
```diff
@@ -105,6 +116,9 @@ def test_orphan_details(self):
         orphan_1 = orphanage[0]
         self.orphan_details_match(orphan_1, tx_child_1, verbosity=1)
+        self.log.info("Checking orphan entry/expiration times")
+        assert_approx(orphan_1["entry"], entry_time, 5)
```
Yeah, if we can, that would be simpler/preference.

Would need the overhead (message sending, processing, orphanage addition, etc.) to always take less than a second, right?  Otherwise there might be intermittent test failure.

If it's always under a second, then maybe something like `setmocktime(int(time.time()))` to force time to start at the beginning of a second might work?

Locally, I very roughly checked runtime and it was consistently under 105ms (not sure how accurate logger is though).  Any good way to benchmark on a less optimistic environment (e.g. congested CI)?

```diff
diff --git a/test/functional/rpc_orphans.py b/test/functional/rpc_orphans.py
index 207647f548..294a73066b 100755
--- a/test/functional/rpc_orphans.py
+++ b/test/functional/rpc_orphans.py
@@ -96,11 +96,14 @@ class OrphanRPCsTest(BitcoinTestFramework):
         tx_child_2 = self.wallet.create_self_transfer(utxo_to_spend=tx_parent_2["new_utxo"])
         peer_1 = node.add_p2p_connection(P2PInterface())
         peer_2 = node.add_p2p_connection(P2PInterface())
+        self.log.info("Setting time to the start of a second")
         entry_time = int(time.time())
+        node.setmocktime(entry_time)
         peer_1.send_and_ping(msg_tx(tx_child_1["tx"]))
         peer_2.send_and_ping(msg_tx(tx_child_2["tx"]))
 
         orphanage = node.getorphantxs(verbosity=2)
+        self.log.info(f"Just after first getorphantxs, entry_time={entry_time}")
         assert tx_in_orphanage(node, tx_child_1["tx"])
         assert tx_in_orphanage(node, tx_child_2["tx"])
```

```
2024-10-07T11:18:24.529000Z TestFramework (INFO): Setting time to the start of a second
2024-10-07T11:18:24.632000Z TestFramework (INFO): Just after first getorphantxs, entry_time=1728299904
```

tdb3 replied:
Thanks.

The following passed.  Will push.

```diff
diff --git a/test/functional/rpc_orphans.py b/test/functional/rpc_orphans.py
index 207647f548..825657a1ed 100755
--- a/test/functional/rpc_orphans.py
+++ b/test/functional/rpc_orphans.py
@@ -96,11 +96,15 @@ class OrphanRPCsTest(BitcoinTestFramework):
         tx_child_2 = self.wallet.create_self_transfer(utxo_to_spend=tx_parent_2["new_utxo"])
         peer_1 = node.add_p2p_connection(P2PInterface())
         peer_2 = node.add_p2p_connection(P2PInterface())
+        self.log.info("Setting time to the start of a second")
         entry_time = int(time.time())
+        node.setmocktime(entry_time)
+        time.sleep(3)
         peer_1.send_and_ping(msg_tx(tx_child_1["tx"]))
         peer_2.send_and_ping(msg_tx(tx_child_2["tx"]))
 
         orphanage = node.getorphantxs(verbosity=2)
+        self.log.info(f"Just after first getorphantxs, entry_time={entry_time}")
         assert tx_in_orphanage(node, tx_child_1["tx"])
         assert tx_in_orphanage(node, tx_child_2["tx"])
 
@@ -117,8 +121,8 @@ class OrphanRPCsTest(BitcoinTestFramework):
         orphan_1 = orphanage[0]
         self.orphan_details_match(orphan_1, tx_child_1, verbosity=1)
         self.log.info("Checking orphan entry/expiration times")
-        assert_approx(orphan_1["entry"], entry_time, 5)
-        assert_approx(orphan_1["expiration"], entry_time + ORPHAN_MAX_RETENTION_TIME, 5)
+        assert_equal(orphan_1["entry"], entry_time)
+        assert_equal(orphan_1["expiration"], entry_time + ORPHAN_MAX_RETENTION_TIME)
 
         self.log.info("Checking orphan details (verbosity 2)")
         orphanage = node.getorphantxs(verbosity=2)
```

tdb3 commented:
Pushed to ae17f9050e0c2a6fd31cd9f5d2f3ad8066f02cc9 to replace `assert_approx` with `assert_equal` (using `setmocktime`)

instagibbs commented on test/functional/test_framework/mempool_util.py:
```diff
@@ -19,6 +19,8 @@
     MiniWallet,
 )
 
+ORPHAN_MAX_RETENTION_TIME = 1200
```
let's name it the same as in the CPP code? `ORPHAN_TX_EXPIRE_TIME`

tdb3 replied:
fixed

instagibbs reviewed (commented):
> Pushed to https://github.com/bitcoin/bitcoin/commit/ae17f9050e0c2a6fd31cd9f5d2f3ad8066f02cc9 to replace assert_approx with assert_equal (using setmocktime)

good call, approx always sets off alarm bells as a test reader :)

we have a chance to outright reject verbosity>2 for extensibility before a release, but I can't think of what other things you'd possible want that can't be given in the 3 modes already included.

concept ACK

itornaza reviewed (approved):
Concept ACK ae17f9050e0c2a6fd31cd9f5d2f3ad8066f02cc9

Enriches the functional tests for the the `getorphantxs` rpc and offers added functionality to conveniently access the expiration times of transactions in the orphanage. 

Broadening the context with refactoring from `rpc_getorphantxs.py` to `rpc_orphans.py` in 469cbac75c088dcb0e3dc5fe8ca7f034148b9487 as a future provision for more orphanage rpcs looks good to me.

rkrux reviewed (approved):
tACK ae17f9050e0c2a6fd31cd9f5d2f3ad8066f02cc9

Successful make and functional tests. Pretty straightforward PR; I agree with @instagibbs's suggestion to keep the variable name in functional test same as in the CPP code.

instagibbs reviewed (changes requested):
I really think `ORPHAN_MAX_RETENTION_TIME` should be called`ORPHAN_TX_EXPIRE_TIME`

after that LGTM

tdb3 commented:
> Would be good to address [#30793 (comment)](https://github.com/bitcoin/bitcoin/pull/30793#discussion_r1788273274) as well. Otherwise there will be another follow-up?

I agree that this is a worthwhile update.  I'll take a deeper look and see if it might fit nicely into this PR (for less code churn).  If not, then it could become its own PR.

> I really think `ORPHAN_MAX_RETENTION_TIME` should be called`ORPHAN_TX_EXPIRE_TIME`

Updated locally.  Will push after looking into the above.

> we have a chance to outright reject verbosity>2 for extensibility before a release, but I can't think of what other things you'd possible want that can't be given in the 3 modes already included.

Is the preference to reject undefined verbosity values for new code?  If so, we can adjust that and any future verbosity levels would be accompanied by updated test(s).

instagibbs commented:
> Is the preference to reject undefined verbosity values for new code (e.g. -1, 5)? If so, we can adjust that and any future verbosity levels would be accompanied by updated test(s).

Rejection now means we can expand usage later if desired, and is more obvious what's happening.

tdb3 committed:
rpc: disallow boolean verbosity in getorphantxs

Updates ParseVerbosity() to support disallowing
boolean verbosity.  Removes boolean verbosity
for getorphantxs to encourage integer verbosity
usage

tdb3 committed:
doc: add rpc guidance for boolean verbosity avoidance

tdb3 committed:
test: explicitly check boolean verbosity is disallowed

tdb3 commented:
Pushed to 38c2ef250817351919eebfbea29549c4f68efac1
- Renamed expiration variable to ORPHAN_TX_EXPIRE_TIME
- getorphantxs now rejects undefined verbosity (<0 and >2), and disallows boolean verbosity through an updated `ParseVerbosity()`.

glozow reviewed (commented):
utACK 0ea84bc362f395fd247623c22942eb5ca3d1b874

rkrux reviewed (approved):
tACK 0ea84bc362f395fd247623c22942eb5ca3d1b874

itornaza reviewed (approved):
tACK 0ea84bc362f395fd247623c22942eb5ca3d1b874

Tested against all of the standard and extended test  harness. Everything looks great!

achow101 commented:
ACK 0ea84bc362f395fd247623c22942eb5ca3d1b874
